File: 301. Remove Invalid Parentheses.js
/**
 * @param {string} s
 * @return {string[]}
 */
var removeInvalidParentheses = function(s) {
    var result = [];
    var l = 0;
    var r = 0;

    for (var i = 0; i < s.length; i++) {
        if (s[i] == '(') {
            l++;
        }
        if (s[i] == ')') {
            if (l > 0) {
                l--;
            } else {
                r++;
            }
        }
    }

    dfs(s, result, l, r, 0);

    return result;
};

var dfs = function (s, result, l, r, index) {
    if ( l == 0 && r == 0 ) {
        if ( isValid(s) ) {
            result.push(s);
        }
        return;
    }

    for (var i=index; i<s.length; i++) {
        if (i != index && s[i] == s[i-1]) {
            continue;
        }

        if (s[i] == '(') {
            var subString = s.substring(0, i) + s.substring(i+1);
            if (l > 0) {
                dfs(subString, result, l-1, r, i);
            }
        }

        if (s[i] == ')'){
            var subString = s.substring(0, i) + s.substring(i+1);
            console.log(subString);
            if (r > 0) {
                dfs(subString, result, l, r-1, i);
            }
        }
    }
};

var isValid = function (s) {
    var count = 0;
    for (var j=0; j<s.length; j++) {
        if (s[j] == '(') {
            count++;
        }
        if (s[j] == ')') {
            count--;
            if (count < 0) {
                return false;
            }
        }
    }

    return count == 0;
};

File: 301. Remove Invalid Parentheses.py
"""
Remove the minimum number of invalid parentheses 
in order to make the input string valid. Return all possible results.

Note: The input string may contain letters other than the parentheses ( and ).

Examples:
"()())()" -> ["()()()", "(())()"]
"(a)())()" -> ["(a)()()", "(a())()"]
")(" -> [""]
"""
class Solution(object):
    def removeInvalidParentheses(self, s):
        self.result = []
        left, right = 0, 0
        for char in s:
            if char == '(':
                left += 1
            if char == ')':
                if left > 0:
                    left -= 1
                else:
                    right += 1
        self.dfs(s, left, right, 0)
        return self.result

    def dfs(self, s, left, right, index):
        if left == 0 and right == 0:
            if self.isValid(s):
                self.result.append(s)
            return
        for i in xrange(index, len(s)):
            if i != index and s[i] == s[i - 1]:
                continue
            if s[i] == '(':
                if left > 0:
                    subString = s[0:i] + s[i + 1:]
                    self.dfs(subString, left - 1, right, i)
            if s[i] == ')':
                if right > 0:
                    subString = s[0:i] + s[i + 1:]
                    self.dfs(subString, left, right - 1, i)

    def isValid(self, s):
        count = 0
        for c in s:
            if c == '(':
                count += 1
            if c == ')':
                count -= 1
                if count < 0:
                    return False
        return count == 0
